packages/core/src/resources/internal-registry.ts

```diff
@@ -0,0 +1,426 @@
+            '.exe',
+            '.dll',
+            '.so',
+            '.dylib',
+            '.bin',
+            '.dat',
+            '.db',
+            '.sqlite',
+            '.jpg',
+            '.jpeg',
+            '.png',
+            '.gif',
+            '.bmp',
+            '.ico',
+            '.tiff',
+            '.webp',
+            '.mp3',
+            '.mp4',
+            '.avi',
+            '.mov',
+            '.wmv',
+            '.flv',
+            '.mkv',
+            '.webm',
+            '.pdf',
+            '.zip',
+            '.tar',
+            '.gz',
+            '.7z',
+            '.rar',
+            '.dmg',
+            '.iso',
+            '.woff',
+            '.woff2',
+            '.ttf',
+            '.otf',
+            '.eot',
+            '.class',
+            '.jar',
+            '.war',
+            '.ear',
+            '.o',
+            '.obj',
+            '.lib',
+            '.a',
+        ];
+        return binaryExtensions.includes(ext);
+    }
+
+    async refresh(): Promise<void> {
+        await this.buildResourceCache();
+    }
+
+    private async buildResourceCache(): Promise<void> {
+        if (!this.config) return;
+
+        this.resourcesCache.clear();
+        this.visitedPaths.clear();
+        this.fileCount = 0;
+
+        const maxFiles = this.config?.maxFiles ?? FileSystemResourceHandler.DEFAULT_MAX_FILES;
+
+        for (const configPath of this.config?.paths ?? []) {
+            if (this.fileCount >= maxFiles) {
+                logger.warn(`Reached maximum file limit (${maxFiles}), stopping scan`);
+                break;
+            }
+
+            try {
+                const root = await fs.realpath(path.resolve(configPath));
+                await this.scanPath(root, 0, root);
+            } catch (error) {
+                logger.warn(
+                    `Failed to scan path '${configPath}': ${error instanceof Error ? error.message : String(error)}`
+                );
+            }
+        }
+
+        logger.debug(
+            `FileSystem resources cached: ${this.resourcesCache.size} resources (${this.fileCount} files scanned)`
+        );
+    }
+
+    private async scanPath(
+        targetPath: string,
+        currentDepth: number,
+        rootBase?: string
+    ): Promise<void> {
+        const resolvedPath = path.resolve(targetPath);
+        let canonical: string;
+        try {
+            canonical = await fs.realpath(resolvedPath);
+        } catch {
+            return;
+        }
+        if (!this.isPathAllowed(canonical)) return;
+
+        const maxDepth = this.config?.maxDepth ?? FileSystemResourceHandler.DEFAULT_MAX_DEPTH;
+        const maxFiles = this.config?.maxFiles ?? FileSystemResourceHandler.DEFAULT_MAX_FILES;
+        const includeHidden = this.config?.includeHidden ?? false;
+        const includeExtensions =
+            this.config?.includeExtensions ?? FileSystemResourceHandler.DEFAULT_INCLUDE_EXTENSIONS;
```

Comment:
set defaults in config zod, don't use defaults here

by this point these things should be set

with this we can likely get rid of the above duplicate constants